### PR TITLE
Make ResourceFormatLoaderGDScript perform loading using cache mode

### LINKED ISSUE #57870: Replace in Files doesn't update file until project is reopened
### Godot version

4.0-alpha2

### System information

Windows 11, Nvidia 1050 Ti, Vulkan

### Issue description

The replace all (no undo) doesn't update the file until the project is reopened

https://user-images.githubusercontent.com/53798244/153260894-827f32d2-11b9-4e36-b42c-02224bc635c2.mp4

Edit:
The video doesn't show for me on some browsers, so I uploaded it:

[video](https://www.youtube.com/watch?v=_wHT9iblIjU)

### Steps to reproduce

1 - Open the replace in files menu with Ctrl + Shift + R
2 - Write the word you want to replace
3 - Write the word you want to replace with
4 - Click on Replace....
5 - Down on the Search Results panel, click on Replace all (no undo)
6 - Nothing happens
7 - Re-open the project, the changes are there

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by volokh0x**
Fixes #57870

Dxo6tBuzUn

### CODE REVIEW

**Comment by vnen** [comment]
I'm not entirely sure this won't give problems since `GDScriptCache` will still have the old version if it was there already. OTOH updating the cache might also become an issue because the original one might be freed and try to remove itself from the cache and end up removing the new one instead (since they are keyed by path).

So, I think at minimum this should be pushed to `GDScriptCache`, just pass the cache mode flag when calling. Then maybe we can have a setting in the scripts themselves to tell if they are in the cache. By default they are, but if there's a replacement the `GDScriptCache` singleton can tell the old one that it's not in cache anymore and therefore it should not try to remove itself once it's freed. Not sure if all this is needed, but it feels safer IMO.

File: modules/gdscript/gdscript_cache.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ class GDScriptCache {
 	static Ref<GDScriptParserRef> get_parser(const String &p_path, GDScriptParserRef::Status status, Error &r_error, const String &p_owner = String());
 	static String get_source_code(const String &p_path);
 	static Ref<GDScript> get_shallow_script(const String &p_path, const String &p_owner = String());
-	static Ref<GDScript> get_full_script(const String &p_path, Error &r_error, const String &p_owner = String());
+	static Ref<GDScript> get_full_script(const String &p_path, Error &r_error, const String &p_owner = String(), ResourceFormatLoader::CacheMode p_cache_mode = ResourceFormatLoader::CACHE_MODE_REUSE);
```
Comment (volokh0x) [inline modules/gdscript/gdscript_cache.h:91]: I assume here that `CACHE_MODE_REUSE` mode was used implicitly before, so this change should be consistent with previous use-cases of `GDScriptCache::get_full_script`

File: modules/gdscript/gdscript_cache.cpp [hunk 2]
```diff
@@ -212,7 +217,13 @@ Ref<GDScript> GDScriptCache::get_full_script(const String &p_path, Error &r_erro
...
 
-	singleton->full_gdscript_cache[p_path] = script.ptr();
+	if (p_cache_mode != ResourceFormatLoader::CACHE_MODE_IGNORE) {
+		singleton->full_gdscript_cache[p_path] = script.ptr();
+		if (!old_script.is_null()) {
+			old_script->cached = false;
+			old_script.unref();
+		}
+	}
 	singleton->shallow_gdscript_cache.erase(p_path);
```
Comment (volokh0x) [inline modules/gdscript/gdscript_cache.cpp:227]: Should we keep it as a side-effect from `script->reload()`?

**Comment by vnen** [comment]
Other than that, I'll reassess this PR soon with some more care to see if it's valuable on its own (even if it's not needed to fix the bug), since I think it's the case

**Comment by Chaosus** [comment]
I've checked and found this PR lead to startup editor crash for my project (I cannot push my project here, but I can post a log):
```
[0] GDScriptFunction::call (C:\GIT\godot\modules\gdscript\gdscript_vm.cpp:464)
[1] GDScript::_super_implicit_constructor (C:\GIT\godot\modules\gdscript\gdscript.cpp:131)
[2] GDScript::_create_instance (C:\GIT\godot\modules\gdscript\gdscript.cpp:157)
[3] GDScript::instance_create (C:\GIT\godot\modules\gdscript\gdscript.cpp:392)
[4] Object::set_script (C:\GIT\godot\core\object\object.cpp:900)
[5] Object::set (C:\GIT\godot\core\object\object.cpp:462)
[6] SceneState::instantiate (C:\GIT\godot\scene\resources\packed_scene.cpp:249)
[7] PackedScene::instantiate (C:\GIT\godot\scene\resources\packed_scene.cpp:1674)
[8] SceneState::instantiate (C:\GIT\godot\scene\resources\packed_scene.cpp:165)
[9] PackedScene::instantiate (C:\GIT\godot\scene\resources\packed_scene.cpp:1674)
[10] EditorNode::load_scene (C:\GIT\godot\editor\editor_node.cpp:3695)
[11] EditorNode::_load_open_scenes_from_config (C:\GIT\godot\editor\editor_node.cpp:4840)
[12] EditorNode::_load_docks (C:\GIT\godot\editor\editor_node.cpp:4630)
[13] EditorNode::_sources_changed (C:\GIT\godot\editor\editor_node.cpp:1037)
[14] CallableCustomMethodPointer<EditorNode,bool>::call (C:\GIT\godot\core\object\callable_method_pointer.h:105)
[15] Callable::call (C:\GIT\godot\core\variant\callable.cpp:51)
[16] Object::emit_signalp (C:\GIT\godot\core\object\object.cpp:1120)
[17] Object::emit_signal<bool> (C:\GIT\godot\core\object\object.h:774)
[18] EditorFileSystem::_notification (C:\GIT\godot\editor\editor_file_system.cpp:1253)
[19] EditorFileSystem::_notificationv (C:\GIT\godot\editor\editor_file_system.h:144)
[20] Object::notification (C:\GIT\godot\core\object\object.cpp:849)
[21] SceneTree::_notify_group_pause (C:\GIT\godot\scene\main\scene_tree.cpp:849)
[22] SceneTree::process (C:\GIT\godot\scene\main\scene_tree.cpp:455)
[23] Main::iteration (C:\GIT\godot\main\main.cpp:2764)
[24] OS_Windows::run (C:\GIT\godot\platform\windows\os_windows.cpp:777)
[25] widechar_main (C:\GIT\godot\platform\windows\godot_windows.cpp:175)
[26] _main (C:\GIT\godot\platform\windows\godot_windows.cpp:199)
[27] main (C:\GIT\godot\platform\windows\godot_windows.cpp:211)
[28] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[29] BaseThreadInitThunk
-- END OF BACKTRACE --
```

**Comment by volokh0x** [comment]
Fixed some merging conflicts.
Meanwhile, this PR was affected by a few others:
1. #58847
Since auto-reloading of scripts is disabled now by default, performing replacement in files makes the editor pop up a window, listing externally modified files and requesting the user to either reload or resave them. I think that reloading them automatically instead is the correct behavior in such case
https://github.com/godotengine/godot/blob/c2eaaef1497f8ad557aea9a14e2b3e0582fcfe54/editor/editor_settings.cpp#L559

2. #62309
Doing `existing->path_cache = String();` in
https://github.com/godotengine/godot/blob/c2eaaef1497f8ad557aea9a14e2b3e0582fcfe54/core/io/resource.cpp#L67-L72
leaves `script` with an empty string filename, and causes an error in subsequent `reload` ing as the result
https://github.com/godotengine/godot/blob/c2eaaef1497f8ad557aea9a14e2b3e0582fcfe54/editor/plugins/script_editor_plugin.cpp#L964-L971

On the other hand, #61870 does not affect this PR directly, since `ScriptEditor::_reload_scripts` takes care of scripts `last_modified_time` in any case

I also tried to get rid of special cache setting by unrefercing the old script before updating cache with a new one

**Comment by Faless** [comment]
The `GDScriptCache` has been changed quite a bit to fix cyclic references, so poking @adamscott to see if he has some feedback here.

**Comment by adamscott** [comment]
I just tested the issue and it seems that the latest changes ended up fixing #57870.

This doesn't necessarily renders this PR moot, just that it needs to be updated to consider the changes that occured.

**Comment by akien-mga** [comment]
Closing as the linked bug is fixed, and this PR has merge conflicts and is stale.

**Pull request closed without merging**